### PR TITLE
MB-62354: Support Cosine Similarity

### PR DESCRIPTION
**Pull request by CascadingRadium**
- Add a Cosine Similarity metric
- Add a method to normalize float32 vectors

### CODE REVIEW

File: vector.go [hunk 1]
```diff
@@ -32,22 +34,37 @@ type VectorField interface {
...
 }
 
+func NormalizeVector(vector []float32) []float32 {
+	// first calculate the magnitude of the vector
+	var mag float32
+	for _, v := range vector {
+		mag += v * v
+	}
+	mag = float32(math.Sqrt(float64(mag)))
+	// zannot normalize a zero vector
```
Comment (metonymic-smokey) [inline vector.go:58]: Nit: "zannot".

Comment (metonymic-smokey) [reply vector.go:58]: Also may be worth adding a check to avoid unnecessarily normalising a unit vector(a vector with magnitude 1).

Comment (CascadingRadium) [reply vector.go]: done macha

File: vector.go [hunk 2]
```diff
@@ -32,22 +34,38 @@ type VectorField interface {
...
 	EuclideanDistance: {},
+	InnerProduct:      {},
 	CosineSimilarity:  {},
 }
 
+func NormalizeVector(vector []float32) []float32 {
+	// first calculate the magnitude of the vector
+	var mag float32
+	for _, v := range vector {
+		mag += v * v
```
Comment (moshaad7) [inline vector.go:55]: v * v can overflow float32

Comment (moshaad7) [reply vector.go:55]: Maybe something like this ? 

```
// Normalize a vector of float32s
func normalize(vec []float32) []float32 {
	var mag float64
	for _, v := range vec {
		mag += float64(v) * float64(v)
	}
	mag = math.Sqrt(mag)

	if mag == 0  ||  mag == 1 {
		return vec
	}

	
	if mag != 0 && mag != 1 {
		for i, v := range vec {
			vec[i] = float32(float64(v) / mag)
		}
	}

	return vec
}
```

Comment (CascadingRadium) [reply vector.go]: thanks macha beast

**Review by abhinavdangeti — CHANGES_REQUESTED** [review]

File: vector.go [hunk 3]
```diff
@@ -32,22 +34,38 @@ type VectorField interface {
...
 const DefaultSimilarityMetric = EuclideanDistance
 
 // Supported similarity metrics for vector fields
 var SupportedSimilarityMetrics = map[string]struct{}{
 	EuclideanDistance: {},
+	InnerProduct:      {},
 	CosineSimilarity:  {},
 }
 
+func NormalizeVector(vector []float32) []float32 {
```
[the remark this answers is not in the document]
Comment (abhinavdangeti) [reply vector.go:51]: Move this to bleve.

**Review by abhinavdangeti — APPROVED** [review]

**Pull request merged**